---
title: "Quality control for PDAC single-cell transcriptomic data"
author: "Mohammadali (Sam) Khalilitousi, Yongjin P. Park"
date: "`r Sys.Date()`"
font_size: 12pt
---

```{r include = FALSE}
source("Util-rmd.R")
data.table::setDTthreads(15)
library(patchwork)
library(mmutilR)
fig.dir <- "Figures/QC"
.mkdir(fig.dir)
```

# Data preprocessing

* FASTQ files for pancreatic ductal adenocarcinoma (PDAC) were obtained from the GSA: `https://ngdc.cncb.ac.cn/gsa/browse/CRA001160`

* Two papers: [Chen _et al._ (2021)](https://pubmed.ncbi.nlm.nih.gov/33819739/) and [Peng _et al._ (2019)](https://pubmed.ncbi.nlm.nih.gov/31273297/)

* FASTQ files were transformed into spliced and unspliced count matrices using kb-python: `https://www.kallistobus.tools/`



```{r results = "asis"}
result.dir <- "Results/QC"
.mkdir(result.dir)
sample.info.file <- result.dir %&% "/Tab_Samples.csv"

if.needed(sample.info.file,
{
    .samples <-
        readxl::read_xlsx("Data/CRA001160.xlsx", 1) %>%
        select(-`Accession`) %>% 
        left_join(readxl::read_xlsx("Data/CRA001160.xlsx", 3)) %>%
        select(`ID`, `Accession`, `Sample name`, `Public description`, `Sex`)
    fwrite(.samples, sample.info.file)
})
```



## 1. Merge all the files into one file set

```{r}
result.dir <- "Results/QC"
.mkdir(result.dir)
merged.hdr <- result.dir %&% "/Merged"
raw.data.dirs <- list.files("Data/processed", pattern = "CRR[0-9]+", full.names = TRUE)
```

We merged `r length(raw.data.dirs)` file sets into one file set, distinguishing the spliced and unspliced features.

```{r}
merged <- fileset.list(merged.hdr)

if(!all(file.exists(unlist(merged)))){

    .hdrs <- raw.data.dirs %&% "/counts_filtered"

    merged <- rcpp_mmutil_merge_file_sets(
        r_headers = .hdrs,
        r_batches = sapply(raw.data.dirs, basename),
        r_mtx = .hdrs %&% ".mtx.gz",
        r_row = .hdrs %&% ".rows.gz",
        r_col = .hdrs %&% ".cols.gz",
        output = merged.hdr)
}
```

Create a mapping between ENSEMBL gene ID and HGNC symbols.

```{r}
features <- fread(merged$row, col.names = "feature", header=FALSE)

parse <- function(s){
    transpose(lapply(strsplit(s, "[_.]"), function(x) c(x[1], x[length(x)])))
}

features[, c("ensembl_gene_id", "status") := parse(`feature`)]

feature.info.file <- result.dir %&% "/feature_info.RDS"

if.needed(feature.info.file,
{

    ensembl <- biomaRt::useMart(biomart='ENSEMBL_MART_ENSEMBL',
                                host='uswest.ensembl.org',
                                path='/biomart/martservice',
                                dataset='hsapiens_gene_ensembl')

    ensembl.hs <- biomaRt::useDataset('hsapiens_gene_ensembl', mart=ensembl)

    .attr <- c('ensembl_gene_id',
               'hgnc_symbol',
               'chromosome_name',
               'transcription_start_site',
               'transcript_start',
               'transcript_end',
               'description',
               'percentage_gene_gc_content')

    .temp <- biomaRt::getBM(attributes=.attr,
                            filters=c('ensembl_gene_id'),
                            values=unique(features$ensembl_gene_id),
                            mart=ensembl.hs,
                            useCache = FALSE)

    .temp <- as.data.table(.temp)

    feature.info <-
        rbind(.temp[transcript_start < transcript_end,
                    .(transcript_start = min(transcript_start),
                      transcript_end = max(transcript_end),
                      percentage_gene_gc_content = mean(percentage_gene_gc_content),
                      chromosome_name = first(chromosome_name),
                      description = first(description)),
                    by = .(ensembl_gene_id, hgnc_symbol)],
              .temp[transcript_start > transcript_end,
                    .(transcript_start = max(transcript_start),
                      transcript_end = min(transcript_end),
                      percentage_gene_gc_content = mean(percentage_gene_gc_content),
                      chromosome_name = first(chromosome_name),
                      description = first(description)),
                    by = .(ensembl_gene_id, hgnc_symbol)])

    saveRDS(feature.info, file=feature.info.file)
})
feature.info <- readRDS(feature.info.file)
```


## 2. Basic Q/C by row-wise and column-wise scores of the combined single-cell RNA-seq data matrix


```{r}
.mt.genes <-
    left_join(features, feature.info) %>%
    filter(chromosome_name == "MT") %>%
    select(feature) %>%
    unlist

.mt.only <- fileset.list(result.dir %&% "/MT")

if.needed(.mt.only,
{
    rcpp_mmutil_copy_selected_rows(merged$mtx,
                                   merged$row,
                                   merged$col,
                                   .mt.genes,
                                   result.dir %&% "/MT");
})
```



```{r}
.mkdir("Table/QC/")
.file <- "Table/QC/cell_scores.csv.gz"
.file.2 <- "Table/QC/feature_scores.csv.gz"

if.needed(c(.file, .file.2),
{

    all.scores <-
        rcpp_mmutil_compute_scores(merged$mtx,
                                   merged$row,
                                   merged$col)

    mt.scores <-
        rcpp_mmutil_compute_scores(.mt.only$mtx,
                                   .mt.only$row,
                                   .mt.only$col)

    score.dt <-
        setDT(all.scores$col) %>%
        left_join(setDT(mt.scores$col),
                  by = "name",
                  suffix = c("", ".MT")) %>%
        as.data.table
    score.dt[is.na(`sum.MT`), `sum.MT` := 0]
    score.dt[, mito := `sum.MT` / `sum` * 100]
    fwrite(score.dt, .file)

    row.score.dt <- setDT(all.scores$row)
    fwrite(row.score.dt, .file.2)
})
score.dt <- fread(.file)
row.score.dt <- fread(.file.2)
```


```{r}
.hist.nnz <- score.dt[, .(.N), by = .(0.1 * ceiling(10 * log10(nnz)))]
names(.hist.nnz) <- c("nnz.log10", "count")
.hist.mito <- score.dt[, .(.N), by = .(ceiling(mito/2) * 2)]
names(.hist.mito) <- c("mito", "count")

.dt <- score.dt %>%
    mutate(nnz.log10.grid = 0.1 * ceiling(10 * log10(nnz)),
           mito.grid = ceiling(mito))

.hist.joint <- .dt[, .(.N), by = .(nnz.log10.grid, mito.grid)]
```

```{r include = FALSE}
MITO.CUTOFF <- 50
NNZ.CUTOFF <- 300
```

#### Select high-quality cells that contain at least $>$ `r NNZ.CUTOFF` non-zero genes expressed with the mitochondrial gene activity $<$ `r MITO.CUTOFF` %


```{r echo = FALSE, results = "asis", fig.width = 5, fig.height=4}
.lab <- function(x) num.sci(10^(x))

p1 <-
    .gg.plot(.hist.nnz, aes(x = `nnz.log10`, y = `count`)) +
    geom_bar(stat="identity") +
    geom_vline(xintercept = log10(NNZ.CUTOFF), colour = "red", lty = 2, size = .5) +
    scale_x_continuous("# non-zero elements per cell", labels = .lab) +
    scale_y_continuous(labels = num.int)

p2 <-
    .gg.plot(.hist.mito, aes(y = `mito`, x = -`count`)) +
    geom_segment(aes(yend = `mito`, xend = 0), size = 1, colour = "gray30") +
    geom_hline(yintercept = MITO.CUTOFF, colour = "red", lty = 2, size = .5) +
    theme(axis.text.x = element_text(angle=45, vjust=1, hjust=1)) +
    ylab("Mitochondrial activity (%)") +
    scale_x_continuous("count", labels = function(x) num.sci(-x))

p3 <-
    .gg.plot(.hist.joint, aes(x = nnz.log10.grid, y = mito.grid, fill = log10(`N`))) +
    theme(panel.background = element_blank()) +
    scale_x_continuous("# non-zero elements per cell", labels = .lab) +
    scale_y_continuous("Mitochondrial activity (%)") +
    geom_tile() +
    geom_vline(xintercept = log10(NNZ.CUTOFF), colour = "red", lty = 2, size = .5) +
    geom_hline(yintercept = MITO.CUTOFF, colour = "red", lty = 2, size = .5) +
    scale_fill_distiller("count", direction = 1, labels = function(x) num.int(10^x)) +
    theme(legend.position = c(1,1), legend.justification = c(1,1))

p0 <- ggplot() + theme_void()

ptop <- (p0|p1) + plot_layout(widths=c(1,3))
pbot <- (p2|p3) + plot_layout(widths=c(1,3))
plt <- (ptop/pbot) +plot_layout(heights=c(1,3))
print(plt)

.file <- str_c(fig.dir, "/Fig_qc_nnz_mito_cutoff.pdf")
.gg.save(.file, plot = plt, width = 5, height = 4)
```

```{r include = FALSE}
ntot.cells <- nrow(score.dt)
nqc.cells <- nrow(score.dt[mito < MITO.CUTOFF & nnz > NNZ.CUTOFF])
```

#### Of `r num.int(ntot.cells)` cells, we retained `r num.int(nqc.cells)` (`r num.round(nqc.cells/ntot.cells*100)` %), discarding `r num.int(ntot.cells - nqc.cells)` cells (`r num.round(100 * (ntot.cells - nqc.cells) / ntot.cells)` %) by applying the Q/C steps

```{r}
cell.qc.data <- fileset.list(result.dir %&% "/cell_qc")
qc.cells <- score.dt[mito < MITO.CUTOFF & nnz > NNZ.CUTOFF, .(name)] %>%
    unlist

if.needed(cell.qc.data,
{
    rcpp_mmutil_copy_selected_columns(merged$mtx,
                                      merged$row,
                                      merged$col,
                                      qc.cells,
                                      output = result.dir %&% "/cell_qc");
})
```

### Cancer-related genes are highly expressed in the scRNA-seq data

```{r}
gwas.data.file <- "gwas_catalog_tidy.tsv.gz"
```

```{r download_full_gwas_catalog, include=FALSE}
gwas.file <- result.dir %&% "/gwas_catalog_v1.0-associations_e105_r2022-02-02.tsv.gz"
if.needed(gwas.file,
{
    url <- "https://www.ebi.ac.uk/gwas/api/search/downloads/full"
    .file <- str_remove(gwas.file, ".gz$")
    download.file(url, destfile = .file)
    R.utils::gzip(.file)
    unlink(.file)
})
if.needed(gwas.data.file,
{
    .dt <-
        fread(gwas.file, sep="\t", quote="") %>%
        select(`MAPPED_GENE`, `DISEASE/TRAIT`, `PVALUE_MLOG`)
    ## remove redundant associations
    .dt <- .dt[order(.dt$PVALUE_MLOG, decreasing = TRUE),
               head(.SD, 1),
               by = .(`MAPPED_GENE`, `DISEASE/TRAIT`)]
    ## remove traits with too few associations
    .count <- .dt[, .(.N), by = .(`DISEASE/TRAIT`)]
    .dt <- left_join(.count[`N` >= 100, ], .dt)[nchar(`MAPPED_GENE`)> 0,]
    ## simply split gene lists and unlist
    .dt <- .dt[,
               .(gene_symbol = unlist(strsplit(`MAPPED_GENE`, split="[ ,.-]+"))),
               by = .(`DISEASE/TRAIT`, PVALUE_MLOG)]
    .dt[, p.value := 10^(-PVALUE_MLOG)]
    fwrite(.dt, file=gwas.data.file)
})
gwas.data <- fread(gwas.data.file)
```

```{r}
.file <- "Table/QC/feature_scores_qc.csv.gz"
if.needed(.file,
{
    all.scores <-
        rcpp_mmutil_compute_scores(cell.qc.data$mtx,
                                   cell.qc.data$row,
                                   cell.qc.data$col)
    .dt <- setDT(all.scores$row)
    fwrite(.dt, .file)
})
feature.scores <- fread(.file) %>%
    rename(`feature` = `name`) %>%
    left_join(features) %>%
    left_join(feature.info) %>%
    as.data.table
```


```{r}
cancer.genes <-
    gwas.data[str_detect(`DISEASE/TRAIT`, "[Cc]ancer"), .(gene_symbol)] %>%
    unique %>%
    unlist

feature.scores[, cancer := "non-CANCER"]
feature.scores[`hgnc_symbol` %in% cancer.genes, cancer := "CANCER"]
feature.scores[str_detect(`description`, "oncogene"), cancer := "CANCER"]               
feature.scores[str_detect(`description`, "tum[ou]+r"), cancer := "CANCER"]
feature.scores[str_detect(`description`, "cancer"), cancer := "CANCER"]
```

```{r echo = FALSE, results = "asis", fig.width = 8, fig.height=2}
.plot.distrib <- function(.aes){
    .gg.plot(feature.scores, .aes) +
        geom_violin(size = 0) +
        theme(legend.position = "none") +
        theme(axis.title.x = element_blank()) +
        ggpubr::stat_compare_means(size = 2) +
        scale_fill_brewer(palette = "Paired") +
        geom_boxplot(outlier.size = 0, outlier.stroke = 0,
                     width = .02, size = .2, fill = "white")
}

.aes <- aes(x = cancer, y = log10(pmax(nnz, 1)), fill = cancer)
p1 <-
    .plot.distrib(.aes) +
    geom_hline(yintercept = 2, lty = 2, colour = "red") +
    scale_y_continuous("# non-zero elements", labels = function(x) num.int(10^(x)))

.aes <- aes(x = cancer, y = log10(pmax(1e-4, `sd`)), fill = cancer)
p2 <-
    .plot.distrib(.aes) +
    scale_y_continuous("max {StdDev, 1e-4}", labels = function(x) 10^(x))

.aes <- aes(x = cancer, y = log10(pmax(1e-4, `mean`)), fill = cancer)
p3 <-
    .plot.distrib(.aes) +
    scale_y_continuous("max {Mean, 1e-4}", labels = function(x) 10^(x))

.aes <- aes(x = cancer, y = log10(pmax(1e-4, `cv`)), fill = cancer)
p4 <-
    .plot.distrib(.aes) +
    scale_y_continuous("max {CV, 1e-4}", labels = function(x) 10^(x))

plt <- p1 | p2 | p3 | p4
print(plt)
.file <- str_c(fig.dir, "/Fig_cancer_vs_noncancer.pdf")
.gg.save(.file, plot = plt, width = 8, height = 2)
```


```{r}
## .gene.dt <-
##     row.score.dt %>%
##     rename(`feature` = `name`) %>%
##     left_join(features) %>%
##     left_join(feature.info) %>%
##     filter(`nnz` >= 100) %>% 
##     as.data.table
##     dcast(ensembl_gene_id ~ status, value.var = "sum", fun.aggregate = sum, fill = 0) %>%
##     filter(spliced > 0, unspliced > 0) %>%
##     as.data.table()
```

