---
title: "Topic modelling for PDAC single-cell transcriptomic data"
author: "Mohammadali (Sam) Khalilitousi, Yongjin P. Park"
date: "`r Sys.Date()`"
font_size: 12pt
---

```{r include = FALSE}
source("Util-rmd.R")
data.table::setDTthreads(15)
library(patchwork)
library(mmutilR)
library(fastTopics)
library(Matrix)

#load("/Users/samtoussi/Desktop/yppQC/fit10_GWAS")
#GWASandBG was Updated 
load("/Users/samtoussi/Desktop/yppQC/fit10_GWASandBG")

fig.dir <- "Figures/Topics"
.mkdir(fig.dir)
QCFolder <- "Results/QC/"
```

# Topic Modeling

## Data Sources

* QC PDAC_GWAS and QC PDAC_GWAS_BG count matrices obtained from the previous step (STEP1_QC_DATA.rmd): `https://causalpathlab.github.io/pdac_velocity_topic/STEP1_QC_Data.html`
* Summary GWAS of invasive pancreatic ductal adenocarcinoma from the JaPAN consortium: `http://www.aichi-med-u.ac.jp/JaPAN/current_initiatives-e.html`

## Data Preparation

### Loading the GWAS data
```{r}
if(file.exists("Results/QC/PDAC_GWAS_BG.mtx.gz")) {
    gwas.bg <- list(mtx = "Results/QC/PDAC_GWAS_BG.mtx.gz",
                      row = "Results/QC/PDAC_GWAS_BG.rows.gz",
                      col = "Results/QC/PDAC_GWAS_BG.cols.gz")
} else {
message("WARNING: You should run STEP 1 of the analysis first to obtain the QC data.")
}
```

### Obtaining ratios of spliced and unspliced counts for topic modeling

```{r, eval=FALSE}
X <- Matrix::readMM(gwas.bg$mtx)
colnames(X) <- read.table(gwas.bg$col)$V1
rownames(X) <- read.table(gwas.bg$row)$V1

s <- X[rownames(X)[str_ends(rownames(X), "_s")],]
u <- X[rownames(X)[str_ends(rownames(X), "_u")],]
#Scaling unspliced counts to have same mean as spliced
u = u * (mean(s)/mean(u))
u = log(u+1)-log(s+1) #u/s ratio
X = rbind(u,-u)
rownames(X) <- str_replace_all(read.table(gwas.bg$row)$V1, c("_s$" = "_u/s", "_u$" = "_s/u"))
X[X<0] = 0 #get rid of negative values
X = X[rowSums(X)>0,colSums(X)>0] #removing rows and columns with sum = 0

X=t(X) #Putting the ratios matrix in the cells x genes format preferred by fastTopics
X = as(X, "dgCMatrix")
```

## Performing topic modeling using fastTopics
```{r, eval=FALSE}
fit10 <- fastTopics::fit_topic_model(X, k = 10, numiter.main = 200, numiter.refine = 200)
save(fit10, ascii=FALSE, file="fit10_GWAS")

fit10DE = fastTopics::diff_count_analysis(fit10,X)
save(fit10DE, ascii=FALSE, file="fit10DE_GWAS")
```

### Viewing topics
```{r, warning=FALSE, message=FALSE}
set.seed(1)
topic_colors10 <- c("gold","forestgreen","dodgerblue","green","skyblue", "pink", "red", "brown", "purple","orange" )

#Labeling cells by their cancer stage
cellNames = rownames(fit10[["L"]])
caseControl = data.frame(labels = matrix("Control", nrow = length(cellNames)), row.names = cellNames)
PDACI = 'CRR241799|CRR034500|CRR034512|CRR034513|CRR034519|CRR034517|CRR034516|CRR241802|CRR034505'
PDACII = 'CRR241798|CRR034499|CRR034501|CRR034504|CRR241803|CRR241804|CRR241800|CRR034511|CRR034506|CRR034510|CRR034509|CRR241801'
PDACIII = 'CRR241805|CRR034503|CRR034507'
caseControl[grep(PDACI,rownames(caseControl),value=TRUE),"labels"] = "PDACI"
caseControl[grep(PDACII,rownames(caseControl),value=TRUE),"labels"] = "PDACII"
caseControl[grep(PDACIII,rownames(caseControl),value=TRUE),"labels"] = "PDACIII"

#view structure plot, grouped by case/control label
fastTopics::structure_plot(fit10,colors = topic_colors10,grouping = caseControl$labels, gap = 25) #view structure plot
```

##Heatmap
```{r, warning=FALSE, message=FALSE}
geneRatioLabels = str_replace_all(read.table("GWASgenes.txt")$V1, c("_s$" = "_u/s", "_u$" = "_s/u"))

#ranking the genes for each topic, by relative "gene ratio" expression levels (high u/s signifies increase in expression, and s/u is opposite)
rankedGenesRatios =  data.frame("k1" = sort(fit10[["F"]][,1], decreasing = T))
rankedGenes = data.frame("k1" = names(sort(fit10[["F"]][,1], decreasing = T)))
newGenes = data.frame("k1" = rankedGenes[!(rankedGenes[,1] %in% geneRatioLabels),1])

for (.x in 1:10) {
    rankedGenes[,paste0("k", .x)] = names(sort(fit10[["F"]][, .x], decreasing = T))
    rankedGenesRatios[,paste0("k", .x)] =  sort(fit10[["F"]][, .x], decreasing = T)
    newGenes[,paste0("k", .x)] = rankedGenes[!(rankedGenes[, .x] %in% geneRatioLabels), .x]
}

gplots::heatmap.2(as.matrix(rankedGenesRatios[1:10,]),dendrogram = "none",Rowv = NA,Colv = NA,labRow = F ,srtCol = 0,trace = "none", density = "none", col=gplots::bluered(100), cellnote =as.matrix(rankedGenes[1:10,]), notecex = 0.5, notecol = "black", colsep=1:10,sepcolor="white",sepwidth=c(0.05,0.05), key.title = "Relative Gene Ratio Value", key.xlab ="Relative Gene Ratio Value" )
```

##Visualizing the structure of the multinomial topic model topic proportions by projection onto a 2-d surface
```{r, warning=FALSE, message=FALSE}
set.seed(1)
umap_plot(fit10)
Y <- umap_from_topics(fit10)
quickplot(Y[,1],Y[,2], color = caseControl$labels) + geom_point(size = 0.005)
```


##Barcharts to visualize the relationship between the loadings (or mixture proportions) and case/control labels
```{r, warning=FALSE, message=FALSE}
set.seed(1)
loadings_plot(fit10, as.factor(caseControl$labels))
```






